File: problems/0347-top-k-frequent-elements/solution.py
# Link: https://leetcode.cn/problems/top-k-frequent-elements/
# Level: Medium
# Date: 2025-05-17

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        cnt = Counter(nums)
        return [item[0] for item in heapq.nlargest(k, cnt.items(), key=lambda x: x[1])]

File: problems/0347-top-k-frequent-elements/solution.cpp
// Link: https://leetcode.cn/problems/top-k-frequent-elements/
// Level: Medium
// Date: 2025-05-17

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freq;
        for (int num : nums) {
            freq[num]++;
        }

        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;
        for (auto& [num, cnt] : freq) {
            minHeap.push({cnt, num});
            if (minHeap.size() > k) {
                minHeap.pop();
            }
        }

        vector<int> res;
        while (!minHeap.empty()) {
            res.push_back(minHeap.top().second);
            minHeap.pop();
        }
        reverse(res.begin(), res.end());
        return res;
    }
};
